fix: Update group dependencies for new rfc model

jennifer-richards opened the pull request:
This eliminates the attempts to get `state.slug` when `state is None`, which superficially fixes the error in ticket #6768.

PR is put up as a draft because I think more attention may be needed - some of the logic identifies RFCs by using the `draft` state. I think that should look at the type instead but I'm still sorting out what's meant to be going on.

Fixes #6768

jennifer-richards committed:
fix: Avoid accessing slug when state is None

jennifer-richards committed:
fix: ignore pre-rfc drafts as rfc-to-rfc references

jennifer-richards committed:
fix: also ignore pre-rfc to pre-rfc refs

jennifer-richards committed:
chore: remove breakpoint

jennifer-richards committed:
fix: ignore subseries non-downrefs

jennifer-richards committed:
fix: label nodes as RFCS when possible

jennifer-richards committed:
fix: get "rfc" flag right

rjsparks commented on ietf/group/views.py:
```diff
@@ -737,26 +737,35 @@ def dependencies(request, acronym, group_type=None):
     references = Q(
         Q(source__group=group) | Q(source__in=cl_docs),
-        source__type="draft",
+        source__type="draft",  # does this need to be type__in=["draft", "rfc"]? Check for informative ref rfc->draft
```
I think likely not, but it _might_ be a change from previous behavior without it.

Since this is only used for groups, we don't have to worry about rfcs that didn't come from drafts here.

The only use such data might have to the consumer of this function is knowing that some RFC inf-reffed this group's active draft. But that's a done deal, so the group won't really be motivated to _do_ anything with the draft given this knowledge. So it's probably noise and shouldn't be on the graph in the first place.

jennifer-richards replied:
I looked at a handful of cases comparing against the 11.16.0 release (i.e., for posterity, the release before the one with the "new" rfc modeling) and did not see any where nodes were missing. That included some selected specifically because they had a `refinfo` relationship pointing from an RFC to a group draft.

Also, I think the only way this could miss a relationship is if there was a relationship from an RFC to a group draft without there also having been a relationship from the draft that _became_ the RFC to the group draft. Does that happen (aside from April 1 RFCs, which I feel pretty comfortable leaving off these plots except perhaps for amusement value)?

jennifer-richards committed:
chore: Remove comment, answer seems to be "no"

rjsparks commented on ietf/group/views.py:
```diff
@@ -737,26 +737,35 @@ def dependencies(request, acronym, group_type=None):
+    both_pre_rfcs = Q(
+        source__type_id="draft", source__states__slug="rfc",
+        target__type_id="draft", target__states__slug="rfc",
+    )
```
it's probably worth adding that you're looking for `states__type="draft"` to all of the places you are looking for `states__slug=="rfc"` in case some other state machine ever gets an `rfc` node.

jennifer-richards replied:
done

rjsparks commented on ietf/group/views.py:
```diff
@@ -771,13 +780,12 @@ def dependencies(request, acronym, group_type=None):
-                "expired": x.get_state("draft").slug == "expired",
-                "replaced": x.get_state("draft").slug == "repl",
-                "group": x.group.acronym if x.group.acronym != "none" else "",
+                "id": x.became_rfc().name if x.became_rfc() else x.name,
```
Check this on a busy group and if the time is noticeable, I think we can get the db to grab this with the primary query instead of n+1 -ing it here.

rjsparks replied:
We might do that in a way the puts the conditional into the query as well, so that we don't see this pattern repeating.

jennifer-richards replied:
On the idr group (see below to gauge whether this is "busy", but it looks it to me), this brute force approach takes the request from 2-2.4 seconds up to 2.2-2.5 seconds on my dev machine. The code here is taking advantage of the caching in `became_rfc()`, so I believe it's at least only paying the N+1 penalty and not hammering the db any worse than that.

<img width="401" alt="image" src="https://github.com/ietf-tools/datatracker/assets/19472766/2c34c700-296d-4d77-a7a2-f28ddc891d02">

jennifer-richards committed:
refactor: Specify state type; drop redundant source doc type check

We're only admitting source__type="draft", so
no need to check it again in filter queries

jennifer-richards committed:
refactor: Specify state type in old code
